### PR TITLE
Update circleci config to circleci workflows 2

### PR DESCRIPTION
**Pull request by impactmass**
Related to #2858, cont of #3618  
Type: **feature|chore**

# Changes
Rewrites CircleCI config.yml to use Workflows v2. It contains the following Workflow jobs:
  - `docker-build`: builds the production bundle image
  - `docker-push`: pushes the production bundle image to Docker Hub. Requires `docker-build`.
  - `deploy-docs`: builds JSDoc files and deploy them to S3. Filtered to master branch only
  - `test`: Runs `reaction test`
  - `eslint`: Runs `npm run lint`

In a later PR, we'll add Snyk job to run Snyk tests. I didn't include that in this PR since the project doesn't have Snyk setup yet.

### CODE REVIEW

**Comment by impactmass** [comment]
@ticean I returned the test job to run outside of docker. It’s a faster build than earlier state.
The only failing job right now is the ESLint. For that I'll just requires updating the code to fix what the lint is catching.
I want to go over the setup again (specifically the save & restore cache part, and maybe decouple test & lint) before removing the WIP on the PR. However, you can throw in comments on it if needed.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,75 +1,217 @@
...
+                  "$DOCKER_REPOSITORY:$CIRCLE_SHA1" \
+                  "$DOCKER_REPOSITORY:%"
+      - run:
+          name: Docker Push
+          command: |
+            docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
+            docker push "$DOCKER_REPOSITORY:$CIRCLE_SHA1"
+            cat docker-cache/docker-tags.txt \
+              | xargs -t -I % \
+                docker push "$DOCKER_REPOSITORY:%"
```
Comment (ticean) [inline .circleci/config.yml]: @impactmass should we add in the condition to prevent Docker push on forks?

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,75 +1,217 @@
-version: 2
+# This CircleCI configuration uses workflows to fan-out to multiple jobs.  The
+# workflow is Dockerized. The first job builds the Docker image which is used
+# in all future steps.
```
Comment (ticean) [inline .circleci/config.yml]: The last couple of lines still accurate?

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,75 +1,217 @@
...
+          command: sudo npm install -g jsdoc
+      - run:
+          name: Build JSDoc files
+          command: |
+            jsdoc . \
+              --verbose \
+              --configure .reaction/jsdoc/jsdoc.json \
+              --readme .reaction/jsdoc/templates/static/README.md
+      - run:
+          name: Deploy Doc files to S3 if on Master
```
Comment (ticean) [inline .circleci/config.yml]: A subtle, non-blocking nitpick... The leftover "if on Master" here indicates (to me) a check within in the step. There's no mention of master in the command as that logic has been moved to the CircleCI workflow. Could this confuse someone on first look?

**Comment by impactmass** [comment]
Updated now to have fork-checking and correct comments.

**Comment by ticean** [comment]
Hi @spencern. @impactmass has the CircleCI workflows ready to go in this PR. Do you think that this can make the cut on 1.9?

* CircleCI builds use workflows (see PR description for more info)
* PR's from forks are not pushed to Docker Hub. Fixes the build failure on forked PR's.
* Linting on the Dockerfile

Here's the [workflow run for this PR](https://circleci.com/workflow-run/2330698f-301a-442e-ac34-3e4cd97d5aa2) if you'd like to see it in action.

**Comment by spencern** [comment]
@ticean @impactmass don't see why not. Let's do it.

**Pull request merged**